Heap Validation

joncham opened the pull request.

lateralusX committed:
Wait on runtime threads to park on joinable thread list during shutdown.

https://github.com/mono/mono/pull/5599 fixed a race condition during shutdown
when runtime threads have come parts of their way through detach, but still
depend on runtime resources, like GC memory. The fix added runtime threads
to the joinable thread list just before they vanished from mono_thread_manage
radar making sure shutdown waited upon the thread before cleaning up.

The above fix slightly changed the behavior of the finalizer thread since it
waits on joinable threads and will now potential block on threads still
executing code (that involves runtime resources). There’s was an assumption
around the threads on the joinable thread list that they should be very close
to complete when added, so join calls coming from the finalizer thread should
almost never block and if it does, the code that remains to execute should not
involve runtime operations risking deadlock situations. Adding the thread to
the list earlier than previously done expose the shutdown to some potential
theoretical problems.

To mitigate the risk and still solve the race condition this commit adds a
mechanism to keep track of active runtime threads until they park on joinable
thread list. The pending counter will be waited upon by the shutdown
thread, just before it does its regular wait on all joinable threads
(after finalizer thread has stopped) to make sure all runtime threads have
been added to the joinable thread list before waiting upon them. Threads are
added to the joinable thread as late as possible, exactly how it’s been done
in the past by sgen_client_thread_detach_with_lock. Shutdown thread will wait
on runtime threads to appear on the list for a short time and if timeout (pending
runtime thread count not reaching 0 before timeout), it will just print a warning
and continue shutdown.

Getting into a wait state during shutdown due to runtime threads not yet added to
joinable threads list should be very rare (hitting previous race condition that was rare),
triggering the timeout should be even more rare, and if that ever happens, we are exposed
to shutdown race condition as we have had in the past, but now we at least get a warning in
the log making it simpler to analyze further.

This commit also fixes a problem with the debugger thread hitting the same race condition as above.
The shutdown thread stopping the debugger thread didn't completely wait for it to stop using runtime
resources before continue shutdown sequence. This triggers the same race condition as when shutting
down regular runtime threads. This commit makes sure stop_debugger_thread waits on the debugger thread
handle to become signaled (happens at the very end of thread lifetime) before continuing the shutdown
logic.

mathieubourgeois committed:
* Generalize commit 0c6932a985175a5cf0cbc59ffd3af4abf3db248e to support LDARG{0|1|2|3}, LDLOC{0|1|2|3}, LDARGS, LDLOCS, LDARG and LDLOC instead of LDLOC and LDLOCS. Improves generated code similar to issue #60945

UnityAlex committed:
Grab part of fix from upstream change: https://github.com/mono/mono/pull/9228 (case 1276200)

joncham committed:
WIP

UnityAlex commented on mono/metadata/boehm-gc.c:
```diff
@@ -946,11 +948,21 @@ mono_gc_free_fixed (void* addr)
 {
 	MonoObject *obj;
 
+	if (validate_heap && (counter++ % validate_frequency) == 0)
```
if you wanted to trim a variable could probably use the frequency to turn heap validation on/off.

Commit:
The NetworkChange event is not supported on Windows (case 1278048)

Provide a proper `PlatformNotSupported` exception for the
`NetworkChange` event on Windows, as it is not currently supported.

UnityAlex committed:
Manual cherrypicking of the following changes since they did not cherrypick cleanly:

https://github.com/Unity-Technologies/mono/commit/7869333c52a2fc2d165effefdbbe97fdf9587b57
https://github.com/mono/mono/pull/9863
https://github.com/Unity-Technologies/mono/commit/f265d9a7838c97d33e8836dfc42d1f946de09041

Needed in order to fix corruption that was occuring to mscorlib's image table on domain reload due to dynamic methods from other images being stored in a hashtable that was not being cleaned up on domain reload.

UnityAlex committed:
Remove hack from mono_free_method. Bypassing cleanup here was causing corruption to persist in mscorlib's method hashtable when a rehash occurred.  Unity profiler may need changes to avoid possessing dangling pointers. Same with the mono profiler (as the comment indicates)

vargaz committed:
[runtime] Throw a MarshalDirectiveException when returning an array from a pinvoke method. (#20425)

Fixes https://github.com/mono/mono/issues/20367.

Some light massaging done by @UnityAlex to cherry pick

UnityAlex committed:
Try to copy monodis into monodistribution/bin dir

ashwinimurt committed:
Use monodis-sgen.exe as default monodis.exe

ashwinimurt committed:
Remove Windows code that does not work

TautvydasZilys committed:
Update Mac toolchain to Xcode 12.2 beta 1.

TautvydasZilys committed:
Use MAP_JIT flag when allocating executable memory on macOS ARM64.

TautvydasZilys committed:
Sign binaries after lipo-ing them since lipo drops signature added by the linker. Also, use JIT entitlement for signing executables.

UnityAlex committed:
Adding API to turn heap validator on and off with a function call.

joncham committed:
Validate field and array elements against type.

joncham committed:
Validate thread local fields.

joncham committed:
Ensure special static slots respect alignment (case 1266322)

Without proper alignment, this may lead to reference types
being stored at non-pointer aligned offsets. Among other
issues this may lead to the GC not scanning those pointers
properly.

UnityAlex committed:
domain will occasionally be null when mono_unity_heap_validation_from_statics is called. Particularly when being called from thread attach.

UnityAlex committed:
Adding mono_validate_object_pointer function which can be called as needed (from unity) to check MonoObjects for corruption.

unity-cla-assistant commented:
[![CLA assistant check](https://cla-assistant.cds.internal.unity3d.com/pull/badge/not_signed)](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/mono?pullRequest=1338) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.cds.internal.unity3d.com/Unity-Technologies/mono?pullRequest=1338) before we can accept your contribution.<br/>**1** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: ashwinimurt<br/>:x: mathieubourgeois<br/>:x: vargaz<br/>:x: lateralusX<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.cds.internal.unity3d.com/check/Unity-Technologies/mono?pullRequest=1338) it.</sub>

UnityAlex commented:
@joncham we can probably close this?
